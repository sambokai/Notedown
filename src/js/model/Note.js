import Persistence from './Persistence';

const localNoteIdCounter = Persistence.readFromLocalStorage('noteIdCounter');
let noteIdCounter = localNoteIdCounter || 0;

const increment = () => {
  noteIdCounter += 1;
  Persistence.writeToLocalStorage('noteIdCounter', noteIdCounter);
  return noteIdCounter;
};

class Note {
  constructor(body) {
    // noinspection JSUnusedGlobalSymbols
    this.body = body;
    // noinspection JSUnusedGlobalSymbols
    this.id = increment();
  }
}

export default Note;
