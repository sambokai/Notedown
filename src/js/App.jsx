/* eslint-disable react/prefer-stateless-function */

import React from 'react';
import { render } from 'react-dom';

import { DragDropContext } from 'react-beautiful-dnd';


import Note from './model/Note';

import TextEditor from './components/TextEditor';
import NoteSelectionList from './components/NoteSelectionList';
import ActionBar from './components/ActionBar';

class App extends React.Component {
  static reorder(list, startIndex, endIndex) {
    const result = Array.from(list);
    const [removed] = result.splice(startIndex, 1);
    result.splice(endIndex, 0, removed);

    return result;
  }

  constructor(props) {
    super(props);

    this.state = {
      notes: [new Note('This is a note.'), new Note('This is a journal entry.')],
      selectedNote: 0,
    };

    this.textEditorTextarea = React.createRef();
  }

  onDragEnd = (result) => {
    // dropped outside the list
    if (!result.destination) {
      return;
    }

    const selectedNoteId = this.state.notes[this.state.selectedNote].id;

    const notes = App.reorder(
      this.state.notes,
      result.source.index,
      result.destination.index,
    );

    this.setState({
      notes,
      selectedNote: notes.findIndex(note => note.id === selectedNoteId),
    });
  };

  handleTextEditorNoteUpdate = (newText) => {
    const notes = this.state.notes.slice();
    notes[this.state.selectedNote].body = newText;
    this.setState({ notes });
  };

  addEmptyNote = () => {
    const newNote = new Note('');

    this.setState({
      notes: this.state.notes.concat([newNote]),
      selectedNote: this.state.notes.length,
    });

    this.focusTextEditorTextarea();
  };

  handleNotesListClick = (noteIndex) => {
    this.setState({ selectedNote: noteIndex });
  };

  dragDropContext(content) {
    return (
      <DragDropContext
        onDragEnd={this.onDragEnd}
      >
        {content}
      </DragDropContext>
    );
  }

  focusTextEditorTextarea() {
    this.textEditorTextarea.current.focus();
  }

  render() {
    const components = {
      app: (
        <div className="container">
          <ActionBar
            onCreateNote={this.addEmptyNote}
          />
          <div className="row py-1 ">
            <div className="col-4">
              <NoteSelectionList
                notes={this.state.notes}
                onSelectNote={this.handleNotesListClick}
                selectedNote={this.state.selectedNote}
              />
            </div>
            <div className="col-8 ">
              <TextEditor
                note={this.state.notes[this.state.selectedNote].body}
                onUpdateNote={this.handleTextEditorNoteUpdate}
                ref={this.textEditorTextarea}
              />
            </div>
          </div>
        </div>),
    };

    return (
      this.dragDropContext(components.app)
    );
  }
}

render(<App />, document.getElementById('app'));
