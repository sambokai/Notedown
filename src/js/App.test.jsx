import React from 'react';

import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import App from './App';
import Note from './model/Note';

const pushMock = jest.fn();
const historyMock = { push: pushMock };

describe('<App/>', () => {
  it('should render correctly, when there are no notes', () => {
    const wrapper = shallow(<App />);

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly, when notes are existing', () => {
    const wrapper = shallow(<App />);
    wrapper.setState({ notes: [{ body: 'This is a note.', id: 5 }] });

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });


  it('should call noteSelectionChanged hook, when the user selects another note by id', () => {
    const firstSelectionId = 99;
    const secondSelectionId = 50;
    const wrapper = shallow(<App match={{ params: { id: firstSelectionId.toString(10) } }} />);

    const selectionChanged = jest.fn();
    wrapper.instance().noteSelectionChanged = selectionChanged;
    wrapper.update();
    wrapper.setProps({ match: { params: { id: secondSelectionId.toString(10) } } });

    expect(selectionChanged)
      .toBeCalledWith(firstSelectionId, secondSelectionId);
  });
});

describe('getNoteIndex()', () => {
  it("should find a note's index in the state array, by id", () => {
    const notes = [
      { id: 60, body: 'This is the first test note.' },
      { id: 99, body: 'This is the second test note.' },
    ];
    const wrapper = shallow(<App />);
    wrapper.setState({ notes });

    expect(wrapper.instance().getNoteIndex(99)).toEqual(1);
  });

  it('should return undefined if the sought-after id, is existing in the note array', () => {
    const notes = [
      { id: 60, body: 'This is the first test note.' },
      { id: 99, body: 'This is the second test note.' },
    ];
    const wrapper = shallow(<App />);
    wrapper.setState({ notes });

    expect(wrapper.instance().getNoteIndex(1000)).toBeUndefined();
  });
});

describe('readSelectedIdFromURL()', () => {
  it('should return null if the id in the url path is not a number', () => {
    const wrapper = shallow(<App match={{ params: { id: 'this-is-a-string' } }} />);

    expect(wrapper.instance().readSelectedIdFromURL()).toBeNull();
  });
});

describe('noteSelectionChanged()', () => {
  it('should delete the empty note if noteSelectionChanged goes from an empty to another note', () => {
    const deleteNoteMock = jest.fn();
    const wrapper = shallow(<App />);
    const notes = [
      new Note('This is Note #1.', 1),
      new Note('', 2),
    ];
    wrapper.setState({ notes });
    wrapper.instance().deleteNote = deleteNoteMock;

    wrapper.instance().noteSelectionChanged(2);

    expect(deleteNoteMock).toBeCalledWith(2, false);
  });
});

describe('addEmptyNote()', () => {
  it('should create an empty note, if creation is allowed', () => {
    const wrapper = shallow(<App history={historyMock} />);
    wrapper.setState({ noteCreationAllowed: true, notes: [] });

    wrapper.instance().addEmptyNote();

    expect(wrapper.state('notes')[0].body).toEqual('');
    expect(wrapper.state('notes')[0].id).toEqual(1);
    expect(wrapper.state('notes')).toHaveLength(1);
  });

  it('should select the newly created note, after creation', () => {
    const wrapper = shallow(<App history={historyMock} />);

    wrapper.instance().addEmptyNote();

    expect(pushMock).toHaveBeenCalledWith('/notes/1');
    expect(pushMock).toHaveBeenCalledTimes(1);
  });
});

describe('deleteNote()', () => {
  it('should delete a note, specified by id', () => {
    const wrapper = shallow(<App />);
    const note = new Note('This is Note #1.', 1);
    wrapper.setState({ notes: [note] });

    wrapper.instance().deleteNote(note.id);

    expect(wrapper.state('notes')).toHaveLength(0);
  });

  it('should, after deletion, select the note, previous to the one deleted', () => {
    const wrapper = shallow(<App history={historyMock} />);
    const notes = [
      new Note('This is Note #1.', 1),
      new Note('This is Note #2', 2),
      new Note('This is Note #3', 3),
      new Note('This is Note #4', 4),
    ];
    wrapper.setState({ notes });

    wrapper.instance().deleteNote(3, true);

    expect(pushMock).toBeCalledWith('/notes/2');
    expect(pushMock).toHaveBeenCalledTimes(1);
  });

  it('should set noteCreationAllowed to true, if the deleted note was empty', () => {
    const wrapper = shallow(<App />);
    const notes = [
      new Note('This is a note with text.', 1),
      new Note('', 2),
    ];
    wrapper.setState({ notes, noteCreationAllowed: false });

    wrapper.instance().deleteNote(2);

    expect(wrapper.state('noteCreationAllowed')).toBe(true);
  });
});

describe('Persistence (Local Storage)', () => {
  describe('syncFromLocalStorage()', () => {
    it('should read noteCreationAllowed from localStorage into app state', () => {
      const wrapper = shallow(<App />);
      wrapper.setState({ noteCreationAllowed: false });

      localStorage.setItem('noteCreationAllowed', 'true');
      wrapper.instance().syncFromLocalStorage();

      expect(wrapper.state('noteCreationAllowed')).toBe(true);
    });

    it('should read notes from localStorage into app state', () => {
      const wrapper = shallow(<App />);
      wrapper.setState({ notes: [] });

      const notes = [
        new Note('This is Note #1.'),
        new Note('This is Note #2.'),
      ];

      localStorage.setItem('notes', JSON.stringify(notes));
      wrapper.instance().syncFromLocalStorage();

      expect(wrapper.state('notes')).toMatchObject(notes);
    });
  });
});

describe('reorder()', () => {
  it(' move an item in a list to the desired index and reorder the remaining array ', () => {
    const initialArray = ['banana', 'apple', 'pear', 'blueberry'];

    const appleFirst = App.reorder(initialArray, 1, 0);
    const bananaSecondLast = App.reorder(initialArray, 0, 2);

    expect(appleFirst).toMatchObject(['apple', 'banana', 'pear', 'blueberry']);
    expect(bananaSecondLast).toMatchObject(['apple', 'pear', 'banana', 'blueberry']);
  });
});
