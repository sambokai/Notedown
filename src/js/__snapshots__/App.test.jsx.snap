// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> should render correctly, when notes are existing 1`] = `
<DragDropContext
  onDragEnd={[Function]}
>
  <div
    className="container"
  >
    <ActionBar
      noteCreationAllowed={true}
      onCreateNote={[Function]}
      onDeleteNote={[Function]}
    />
    <div
      className="row py-1 no-gutters"
      style={
        Object {
          "height": "450px",
        }
      }
    >
      <NoteSelector
        notes={
          Array [
            Object {
              "body": "This is a note.",
              "id": 5,
            },
          ]
        }
        selectedId={null}
      />
      <div
        className="col-8 pl-1"
      >
        <TextEditor
          noNoteMessage="Please select a note to display it."
          note={Object {}}
          onUpdateNote={[Function]}
          placeholder="Write your notes in here..."
        />
      </div>
    </div>
  </div>
</DragDropContext>
`;

exports[`<App/> should render correctly, when there are no notes 1`] = `
<DragDropContext
  onDragEnd={[Function]}
>
  <div
    className="container"
  >
    <ActionBar
      noteCreationAllowed={true}
      onCreateNote={[Function]}
      onDeleteNote={[Function]}
    />
    <div
      className="row py-1 no-gutters"
      style={
        Object {
          "height": "450px",
        }
      }
    >
      <NoteSelector
        notes={Array []}
        selectedId={null}
      />
      <div
        className="col-8 pl-1"
      >
        <TextEditor
          noNoteMessage="Create a new note by clicking the 'New Note' Button"
          note={Object {}}
          onUpdateNote={[Function]}
          placeholder="Write your notes in here..."
        />
      </div>
    </div>
  </div>
</DragDropContext>
`;
